include/xparrow/xbuffer.hpp

```diff
@@ -0,0 +1,239 @@
+        using size_type = typename base_type::size_type;
+
+        xbuffer() = default;
+        explicit xbuffer(size_type size);
+        xbuffer(pointer data, size_type size);
+        
+        ~xbuffer();
+        
+        xbuffer(const xbuffer&);
+        xbuffer& operator=(const xbuffer&);
+
+        xbuffer(xbuffer&&);
+        xbuffer& operator=(xbuffer&&);
+
+        using base_type::empty;
+        using base_type::size;
+        using base_type::data;
+
+        void resize(size_type new_size);
+
+        void swap(xbuffer&) noexcept;
+        bool equal(const xbuffer& rhs) const;
+
+    private:
+
+        pointer allocate(size_type size) const;
+        void deallocate(pointer mem) const;
+    };
+
+    template <class T>
+    bool operator==(const xbuffer<T>& lhs, const xbuffer<T>& rhs);
+
+    /**
+     * @class xbuffer_reference
+     * @brief Object that wraps a piece of contiguous memory
+     * but does not own it.
+     */
+    /*class xbuffer_reference : private xbuffer_data
+    {
+    public:
+
+        using base_type = xbuffer_data;
+
+        xbuffer_reference(uint8_t* data, std::size_t size);
+
+        using base_type::size;
+        using base_type::data;
+        
+        void swap(xbuffer_reference&) noexcept;
+    };*/
+
+    /******************************
+     * xbuffer_data implementation *
+     ******************************/
+
+    namespace impl
+    {
+        template <class T>
+        bool xbuffer_data<T>::empty() const noexcept
+        {
+            return size() == size_type(0);
+        }
+
+        template <class T>
+        auto xbuffer_data<T>::size() const noexcept -> size_type
+        {
+            return m_size;
+        }
+
+        template <class T>
+        template <class U>
+        U* xbuffer_data<T>::data() noexcept
+        {
+            return reinterpret_cast<U*>(p_data);
+        }
+
+        template <class T>
+        template <class U>
+        const U* xbuffer_data<T>::data() const noexcept
+        {
+            return reinterpret_cast<const U*>(p_data);
+        }
+
+        template <class T>
+        void xbuffer_data<T>::swap(xbuffer_data<T>& rhs) noexcept
+        {
+            std::swap(p_data, rhs.p_data);
+            std::swap(m_size, rhs.m_size);
+        }
+
+        template <class T>
+        bool xbuffer_data<T>::equal(const xbuffer_data<T>& rhs) const
+        {
+            return m_size == rhs.m_size && std::equal(p_data, p_data + m_size, rhs.p_data);
+        }
+    }
+
+    /*************************
+     * xbuffer implementation *
+     *************************/
+
+    template <class T>
+    xbuffer<T>::xbuffer(size_type size)
+        : base_type{allocate(size), size}
+    {
+    }
+
+    template <class T>
+    xbuffer<T>::xbuffer(pointer data, size_type size)
+        : base_type{data, size}
+    {
+    }
+        
+    template <class T>
+    xbuffer<T>::~xbuffer()
+    {
+        deallocate(this->p_data);
+        this->p_data = nullptr;
```

Comment:
Arent these 2 lines unnecessary at this point?

Reply:
For the moment I agree that they are not mandatory.